src/python/pants/backend/docker/goals/package_image.py

```diff
@@ -187,6 +210,82 @@ def get_context_root(self, default_context_root: str) -> str:
         return os.path.normpath(context_root)
 
 
+@dataclass(frozen=True)
+class ImageRefRegistry:
+    registry: DockerRegistryOptions | None
+    repository: str
+    tags: tuple[ImageRefTag, ...]
+
+
+@dataclass(frozen=True)
+class ImageRefTag:
+    template: str
+    formatted: str
+    full_name: str
+
+
+@dataclass(frozen=True)
+class DockerInfoV1:
+    """The format of the .docker-info.json file."""
+
+    version: Literal[1]
+    image_id: str
+    # It'd be good to include the digest here (e.g. to allow 'docker run
+    # registry/repository@digest'), but that is only known after pushing to a V2 registry
```

Comment:
I don't think this is accurate. The `image_id` on the `BuiltDockerImage` is the digest.

Reply:
I find it very confusing, but I think docker has two hashes of an image (image ID and digest). For instance, https://stackoverflow.com/q/39811230/1256624 and https://stackoverflow.com/q/56364643/1256624. 🤯 

Summary of my understanding of the differences:

| | digest | image ID |
|---|---|---|
| output for a `DOCKER_BUILDKIT=1` build | ❌ | ✅ |
| output for a `DOCKER_BUILDKIT=0` build |❌ | ✅ truncated |
| available for a built image (before pushing) | ❌ | ✅ |
| available for an pulled/pushed image (registry V2 only) | ✅ | ✅ |
| refer to image via `registry/repository@sha256:abcdef...` | ✅ pulls if required | ❌ |
| refer to image via `sha256:abcdef...` or `abcdef...` | ❌ | ✅ local images only |

For instance (output as `#>` comments):

```shell
docker pull python:3.9-slim
#> ...
#> Digest: sha256:f38c381db444c685b86325ad7feb9dafcacd1ef2d96766bb08235644383194d5
#> ...

# inspect the info:
docker images --digests --no-trunc python
#> REPOSITORY   TAG        DIGEST                                                                    IMAGE ID                                                                  CREATED         SIZE
#> ...
#> python       3.9-slim   sha256:f38c381db444c685b86325ad7feb9dafcacd1ef2d96766bb08235644383194d5   sha256:1a8db82bc51490fbba59314b0032f0086983b6875d1d09c8d660fdcf67fc677e   8 days ago      118MB
#> ...

# note: image ID is different to digest
# (strangely `docker images --digests --no-trunc python:3.9-slim` doesn't list the digest 🤷 )

# run using the digest with @:
docker run python@sha256:f38c381db444c685b86325ad7feb9dafcacd1ef2d96766bb08235644383194d5 echo worked
#> worked

# run using the image ID with @:
docker run python@sha256:1a8db82bc51490fbba59314b0032f0086983b6875d1d09c8d660fdcf67fc677e echo worked
#> Unable to find image 'python@sha256:1a8db82bc51490fbba59314b0032f0086983b6875d1d09c8d660fdcf67fc677e' locally
#> docker: Error response from daemon: manifest for python@sha256:1a8db82bc51490fbba59314b0032f0086983b6875d1d09c8d660fdcf67fc677e not found: manifest unknown: manifest unknown.

# run using the image ID directly:
docker run sha256:1a8db82bc51490fbba59314b0032f0086983b6875d1d09c8d660fdcf67fc677e echo worked
#> worked

# (can also be truncated and/or without sha256: prefix, e.g. `docker run sha256:1a8db82bc ...` or  `docker run 1a8db82bc ...`)

# run using the digest directly:
docker run sha256:f38c381db444c685b86325ad7feb9dafcacd1ef2d96766bb08235644383194d5 echo worked
#> docker: Error response from daemon: No such image: sha256:f38c381db444c685b86325ad7feb9dafcacd1ef2d96766bb08235644383194d5.
```

`docker build` outputs the image ID (with buildkit, the full image ID; without it, a truncated image ID), and the digest seems to only exist for images that are pulled or pushed.

Reply:
Oh wow. OK, thanks for that clarification.